### ISSUE TITLE
False-Positive: type sometimes not deduced correctly when using `Union` and `isinstance`

### ISSUE THREAD

**Issue by Kwasniok** [python/mypy#11070]
**Bug Report**

mypy cannot correctly deduce the type of an object if two unions of types are involved - even though  it correctly does so if only one union is involved.

**To Reproduce**
```python
# test.py
from typing import Union

class A:
    def a(self) -> bool:
        return True
class B:
    def b(self) -> bool:
        return True

def f(x:Union[A,B], y:Union[A,B]) -> bool:
    if isinstance(x,A) and isinstance(y,A):
        return x.a() and y.a()
    if isinstance(x,A) and isinstance(y,B):
        return x.a() and y.b()
    if isinstance(x,B) and isinstance(y,A):
        return x.b() and y.a()
    return x.b() and y.b()  # <- here

def g(x:Union[A,B]) -> bool:
    if isinstance(x,A):
        return x.a()
    return x.b() # <- but not here
```
Using `python -m mypy --show-error-codes test.py`

**Expected Behavior**

mypy should not find any issues.

**Actual Behavior**

```
test.py:18: error: Item "A" of "Union[A, B]" has no attribute "b"  [union-attr]
```

Observe how there is no issue for function `g`.

**Your Environment**

- Mypy version used: `mypy 0.812`
- Mypy command-line flags: `--show-error-codes` (happens also without this flag)
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: `Python 3.9.6`
- Operating system and version: `Linux version 5.13.10-arch1-1 (linux@archlinux) (gcc (GCC) 11.1.0, GNU ld (GNU Binutils) 2.36.1)`

**Possibly Related to**
#10136

**Comment by erictraut** [comment]
Mypy cannot perform type narrowing in this case because of the way you have structured the code. Doing so would require mypy to track types of variables that are conditioned on the type of another variables. For example, it would need to deduce that "if the type of `y` is `A`, then the type of `x` is `A`, but otherwise the type of `x` is `A | B`. This type of conditional type tracking would be very complex and computationally expensive. I'm not aware of any static type checker in any language that does this.

The good news is that you can make a small change to the logic in function `f`, and mypy will be able to perform type narrowing.

```python
def f(x: Union[A, B], y: Union[A, B]) -> bool:
    if isinstance(x, A):
        if isinstance(y, A):
            return x.a() and y.a()
        if isinstance(y, B):
            return x.a() and y.b()

    if isinstance(x, B):
        if isinstance(y, A):
            return x.b() and y.a()

    return x.b() and y.b()
```

**Comment by Kwasniok** [comment]
Thank you @erictraut ! Your proposed solution is exactly what I needed. I was not aware of the complexity of the underlying problem here. And thank you for the detailed answer. It helped me to understand why your solution works so nicely.
